README.md
Pcompress
=========

Copyright (C) 2012-2013 Moinak Ghosh. All rights reserved.
Use is subject to license terms.
moinakg (_at) gma1l _dot com.
Comments, suggestions, code, rants etc are welcome.

Pcompress is an archiver that also does compression and decompression in
parallel by splitting input data into chunks. It has a modular structure
and includes support for multiple algorithms like LZMA, Bzip2, PPMD, etc,
with SKEIN/SHA checksums for data integrity. Compression algorithms are
selected based on the file type to maximize compression gains using a file
and data anaylis based adaptive technique. It also includes various data
transformation filters to improve compression.

It also implements Variable Block Deduplication and Delta Compression
features based on a Polynomial Fingerprinting scheme. Delta Compression
is done via the widely popular bsdiff algorithm. Similarity is detected
using a technique based on MinHashing. Deduplication metadata is also
compressed to reduce overheads. In addition to all these it can internally
split chunks at file and rabin boundaries to help Dedupe and compression.

It has low metadata overhead and overlaps I/O and compression to achieve
maximum parallelism. It also bundles a simple slab allocator to speed
repeated allocation of similar chunks. It can work in pipe mode, reading
from stdin and writing to stdout. SIMD vector optimizations using the x86
SSE instruction set are used to speed up various operations. Finally it
supports 14 compression levels to allow for ultra compression parameters
in some algorithms.

Pcompress also supports encryption via AES, Salsa20 and uses Scrypt from
Tarsnap for Password Based Key generation. A unique key is generated per
session even if the same password is used and HMAC is used to do authentication.

Links of Interest
=================

Project Home Page: http://moinakg.github.io/pcompress/

http://moinakg.github.io/pcompress/#deduplication-chunking-analysis

http://moinakg.github.io/pcompress/#compression-benchmarks

http://moinakg.wordpress.com/2013/04/26/pcompress-2-0-with-global-deduplication/

http://moinakg.wordpress.com/2013/03/26/coordinated-parallelism-using-semaphores/

http://moinakg.wordpress.com/2013/06/11/architecture-for-a-deduplicated-archival-store-part-1/

http://moinakg.wordpress.com/2013/06/15/architecture-for-a-deduplicated-archival-store-part-2/

Standard Usage
==============
    Standard usage consists of a few common options to control basic behavior. A variety of
    parameters including global deduplication are automatically set based on the compression
    level.

    Archiving
    ---------
       pcompress -a [-v] [-l <compress level>] [-s <chunk size>] [-c <algorithm>]
                    [<file1> <directory1> <file2> ...] [-t <number>] [-S <chunk checksum>]
                    <archive filename or '-'>

       Archives a given set of files and/or directories into a compressed PAX archive. The
       PAX datastream is encoded into a custom format compressed file that can only be
       handled by Pcompress.

       -a       Enables archive mode where pathnames specified in the command line are
                archived using LibArchive and then compressed.

       -l <compress level>
                Select a compression level from 1 (least compression, fastest) to 14
                (ultra compression, slow). Default: 6

       -s <chunk size>
                Archive data is split into chunks that are processed in parallel. This value
                specifies the maximum chunk size. Blocks may be smaller than this. Values
                can be in bytes or <number><suffix> format where suffix can be k - KB, m - MB,
                g - GB. Default: 8m
                Larger chunks can produce better compression at the cost of memory.

       -c <algorithm>
                Specifies the compression algorithm to use. Default algorithm when archiving
                is adapt2 (Second Adaptive Mode). This is the ideal mode for archiving giving
                best compression gains. However adapt (Adaptive Mode) can be used which is a
                little faster but give lower compression gains.
                Other algorithms can be used if all the files are of the same known type. For
                example ppmd (slow) or libbsc (fast) can be used if all the files only have
                ASCII text. See section "Compression Algorithms" for details.

       -v       Enables verbose mode where each file/directory is printed as it is processed.

       -t <number>
                Sets the number of threads that Pcompress can use. Pcompress automatically
                uses thread count = core count. However with larger chunk size (-s option)
                and/or ultra compression levels, large amounts of memory can be used. In this
                case thread count can be reduced to reduce memory consumption.

       -S <chunk checksum>
                Specify then chunk checksum to use. Default: BLAKE256. The following checksums
                are available:

                     CRC64 - Extremely Fast 64-bit CRC from LZMA SDK.
                    SHA256 - SHA512/256 version of Intel's optimized (SSE,AVX) SHA2 for x86.
                    SHA512 - SHA512 version of Intel's optimized (SSE,AVX) SHA2 for x86.
                 KECCAK256 - Official 256-bit NIST SHA3 optimized implementation.
                 KECCAK512 - Official 512-bit NIST SHA3 optimized implementation.
                  BLAKE256 - Very fast 256-bit BLAKE2, derived from the NIST SHA3
                             runner-up BLAKE.
                  BLAKE512 - Very fast 256-bit BLAKE2, derived from the NIST SHA3
                             runner-up BLAKE.

                 The fastest checksum is the BLAKE2 family.

       <archive filename>
                Pathname of the resulting archive. A '.pz' extension is automatically added
                if not already present. This can also be specified as '-' in order to send
                the compressed archive stream to stdout.

    Single File Compression
    -----------------------
       pcompress -c <algorithm> [-l <compress level>] [-s <chunk size>] [-p] [<file>]
                 [-t <number>] [-S <chunk checksum>] [<target file or '-'>]

       Takes a single file as input and produces a compressed file. Archiving is not performed.
       This can also work in streaming mode.

       -c <algorithm>
                See above. Also see section "Compression Algorithms" for details.

       -l <compress level>
       -s <chunk size>
       -t <number>
       -S <chunk checksum>
                See above.
                Note: In singe file compression mode with adapt2 or adapt algorithm, larger
                      chunks may not produce better compression. Smaller chunks can result
                      in better data analysis here.

       -p       Make Pcompress work in streaming mode. Data is ingested via stdin
                compressed and output via stdout. No filenames are used.

       <target file>
                Pathname of the compressed file to be created. This can be '-' to send the
                compressed data to stdout.

    Decompression and Archive extraction
    ------------------------------------
       pcompress -d <compressed file or '-'> [-m] [-K] [-i] [<target file or directory>]

       -m        Enable restoring *all* permissions, ACLs, Extended Attributes etc.
                 Equivalent to the '-p' option in tar. Ownership is only extracted if run as
                 root user.
       -K        Do not overwrite newer files.
       -i        Only list contents of the archive, do not extract.

       -m and -K are only meaningful if the compressed file is an archive. For single file
       compressed mode these options are ignored.

       <compressed file>
                Specifies the compressed file or archive. This can be '-' to indicate reading
                from stdin while write goes to <target file>

       <target file or directory>
                This can be a filename or a directory depending on how the archive was created.
                If single file compression was used then this can be the name of the target
                file that will hold the uncompressed data.
                If this is omitted then an output file is created by appending '.out' to the
                compressed filename.

                If Archiving was done then this should be the name of a directory into which
                extracted files are restored. The directory is created if it does not exist.
                If this is omitted the files are extracted into the current directory.

Compression Algorithms
======================

    lzfx    - Fast, average compression. At high compression levels this can be faster
              than LZ4.
              Effective Levels: 1 - 5
    lz4     - Very Fast, sometimes better compression than LZFX.
              Effective Levels: 1 - 3
    zlib    - Fast, better compression.
              Effective Levels: 1 - 9
    bzip2   - Slow, much better compression than Zlib.
              Effective Levels: 1 - 9

    lzma    - Very slow. Extreme compression. Recommended: Use lzmaMt variant mentioned
              below.
              Effective Levels: 1 - 14
              Till level 9 it is standard LZMA parameters. Levels 10 - 12 use
              more memory and higher match iterations so are slower. Levels
              13 and 14 use larger dictionaries upto 256MB and really suck up
              RAM. Use these levels only if you have at the minimum 4GB RAM on
              your system.
    lzmaMt  - This is the multithreaded variant of lzma and typically runs faster.
              However in a few cases this can produce slightly lesser compression
              gain.

    PPMD    - Slow. Extreme compression for Text, average compression for binary.
              In addition PPMD decompression time is also high for large chunks.
              This requires lots of RAM similar to LZMA. PPMd requires
              at least 64MB X core-count more memory than the other modes.
              Effective Levels: 1 - 14.

    Adapt   - Synthetic mode with text/binary detection. For pure text data PPMD is
              used otherwise Bzip2 is selected per chunk.
              Effective Levels: 1 - 14
    Adapt2  - Slower synthetic mode. For pure text data PPMD is otherwise LZMA is
              applied. Can give very good compression ratio when splitting file
              into multiple chunks.
              Effective Levels: 1 - 14
              Since both LZMA and PPMD are used together memory requirements are
              large especially if you are also using extreme levels above 10. For
              example with 100MB chunks, Level 14, 2 threads and with or without
              dedupe, it uses upto 2.5GB physical RAM (RSS).

    none    - No compression. This is only meaningful with -G or -D. So Dedupe
              can be done for post-processing with an external utility.

Enabled features based on Compression Level
===========================================

    1 to 3  - No features, just compression and archiving, if needed.
    4       - Global Deduplication with avg block size of 8KB.
    5       - Global Dedup block size 8KB, Adaptive Delta Encoding.
    6 to 8  - Global Dedup block size reduced to 4KB, Adaptive Delta Encoding.
    9       - Global Dedup block size reduced to 2KB, Adaptive Delta Encoding, Dispack.
    10      - Global Dedup block size 2KB, Adaptive Delta Encoding with extra rounds, Dispack,
              LZP Preprocessing
    10 - 14 - Global Dedup block size 2KB, Adaptive Delta Encoding with extra rounds, Dispack,
              LZP Preprocessing, PackJPG filter for Jpegs.

Encryption
==========
    Pcompress supports encryption and authentication in both archive and single-file
    compresion modes. Encryption options are discussed below.

    NOTE: When using pipe-mode via -p the only way to provide a password is to use '-w'.
          See below.

       -e <ALGO>
                Encrypt chunks using the given encryption algorithm. The algo parameter
                can be one of AES or SALSA20. Both are used in CTR stream encryption
                mode.
                The password can be prompted from the user or read from a file. Unique
                keys are generated every time pcompress is run even when giving the same
                password. Of course enough info is stored in the compresse file so that
                the key used for the file can be re-created given the correct password.

                Default key length if 256 bits but can be reduced to 128 bits using the
                '-k' option.

                The Scrypt algorithm from Tarsnap is used
                (See: http://www.tarsnap.com/scrypt.html) for generating keys from
                passwords. The CTR mode AES mechanism from Tarsnap is also utilized.

       -w <pathname>
                Provide a file which contains the encryption password. This file must
                be readable and writable since it is zeroed out after the password is
                read.

       -k <key length>
                Specify the key length. Can be 16 for 128 bit keys or 32 for 256 bit
                keys. Default value is 32 for 256 bit keys.


Advanced usage
==============
    A variety of advanced options are provided if one wishes fine grained control
    as opposed to automatic settings. If advanced options are used then auto-setting
    of parameters get disabled. The various advanced options are discussed below.

    Chunk-level Deduplication
    -------------------------
    Attempt Polynomial fingerprinting based deduplication on a per-chunk basis:
       pcompress -D ...

    Perform Delta Encoding in addition to Identical Dedup:
       pcompress -E ... - This also implies '-D'. This performs Delta Compression
                          between 2 blocks if they are 40% to 60% similar. The
                          similarity %age is selected based on the dedupe block
                          size to balance performance and effectiveness.
       pcompress -EE .. - This causes Delta Compression to happen if 2 blocks are
                          at least 40% similar regardless of block size. This can
                          effect greater final compression ratio at the cost of
                          higher processing overhead.

       -F       Perform Fixed Block Deduplication. This is faster than fingerprinting
                based content-aware deduplication in some cases. However this is mostly
                usable for disk dumps especially virtual machine images. This generally
                gives lower dedupe ratio than content-aware dedupe (-D) and does not
                support delta compression.

    Global Deduplication
    --------------------
       -G       This flag enables Global Deduplication. This makes pcompress maintain an
                in-memory index to lookup cryptographic block hashes for duplicates. Once
                a duplicate is found it is replaced with a reference to the original block.
                This allows detecting and eliminating duplicate blocks across the entire
                dataset. In contrast using only '-D' or '-F' flags does deduplication only
                within the chunk but uses less memory and is much faster than Global Dedupe.

                The '-G' flag can be combined with either '-D' or '-F' flags to indicate
                rabin chunking or fixed chunking respectively. If these flags are not
                specified then the default is to assume rabin chunking via '-D'.
                All other Dedupe flags have the same meanings in this context.

                Delta Encoding is not supported with Global Deduplication at this time. The
                in-memory hashtable index can use upto 75% of free RAM depending on the size
                of the dataset. In Pipe mode the index will always use 75% of free RAM since
                the dataset size is not known. This is the simple full block index mode. If
                the available RAM is not enough to hold all block checksums then older block
                entries are discarded automatically from the matching hash slots.

                If pipe mode is not used and the given dataset is a file then Pcompress
                checks whether the index size will exceed three times of 75% of the available
                free RAM. In such a case it automatically switches to a Segmented Deduplication
                mode. Here data is first split into blocks as above. Then upto 2048 blocks are
                grouped together to form a larger segment. The individual block hashes for a
                segment are stored on a tempfile on disk. A few min-values hashes are then
                computed from the block hashes of the segment which are then loaded into the
                index. These hashes are used to detect segments that are approximately similar
                to each other. Once found the block hashes of the matching segments are loaded
                from the temp file and actual deduplication is performed. This allows the
                in-memory index size to be approximately 0.0025% of the total dataset size and
                requires very few disk reads for every 2048 blocks processed.
                  
                In pipe mode Global Deduplication always uses a segmented similarity based
                index. It allows efficient network transfer of large data.
                  
       -B <0..5>
                Specify an average Dedupe block size. 0 - 2K, 1 - 4K, 2 - 8K ... 5 - 64K.
                Default deduplication block size is 4KB for Global Deduplication and 2KB
                otherwise.
       -B 0
                This uses blocks as small as 2KB for deduplication. This option can be
                used for datasets of a few GBs to a few hundred TBs in size depending on
                available RAM.

       -L       Enable LZP pre-compression. This improves compression ratio of all
                algorithms with some extra CPU and very low RAM overhead. Using
                delta encoding in conjunction with this may not always be beneficial.
                However Adaptive Delta Encoding is beneficial along with this.

       -P       Enable Adaptive Delta Encoding. It can improve compresion ratio further
                for data containing tables of numerical values especially if those are
                in an arithmetic series. In this implementation basic Delta Encoding is
                combined with Run-Length encoding and Matrix transpose
       NOTE -   Both -L and -P can be used together to give maximum benefit on most
                datasets.

       -x       Perform Dispack Encoding. This is useful to translate x86 call and jmp
                relative offsets to absolute values which compress better. The given
                chunk is split into 32KB blocks and some heuristics are used per block
                to identify whether it represents x86 instruction stream or not. This
                works only when archiving.

       -j       Enable PackJPG processing for Jpeg files. This works only when archiving.

       -M       Display memory allocator statistics.
       -C       Display compression statistics.

Environment Variables
=====================

    Set ALLOCATOR_BYPASS=1 in the environment to avoid using the the built-in
    allocator. Due to the the way it rounds up an allocation request to the nearest
    slab the built-in allocator can allocate extra unused memory. In addition you
    may want to use a different allocator in your environment.

    The variable PCOMPRESS_INDEX_MEM can be set to limit memory used by the Global
    Deduplication Index. The number specified is in multiples of a megabyte.

    The variable PCOMPRESS_CACHE_DIR can point to a directory where some temporary
    files relating to the Global Deduplication process can be stored. This for example
    can be a directory on a Solid State Drive to speed up Global Deduplication. The
    space used in this directory is proportional to the size of the dataset being
    processed and is slightly more than 8KB for every 1MB of data.

    The default checksum used for block hashes during Global Deduplication is SHA256.
    However this can be changed by setting the PCOMPRESS_CHUNK_HASH_GLOBAL environment
    variable. The list of allowed checksums for this is:

    SHA256   , SHA512
    KECCAK256, KECCAK512
    BLAKE256 , BLAKE512
    SKEIN256 , SKEIN512

    Even though SKEIN is not supported as a chunk checksum (not deemed necessary
    because BLAKE2 is available) it can be used as a dedupe block checksum. One may
    ask why? The reasoning is we depend on hashes to find duplicate blocks. Now SHA256

Changelog
== 3.1 Bugfix Release ==
Avoid auto-selection variable chunking for buffer sizes below threshold.
Do not auto-select Global Dedupe for below threshold buffers.
Fix issue #18.
Do not try to generate a target filename in pipe mode.
Scale default compression buffer size for levels > 8.
Fix issue #17.
Use LZ4 and Libbsc extra padding space for the compression buffer in adaptive modes.
Fix issue #16.
Fix issue #15.
Fix issue #14
Updated fix for issue #12
Add check for correct Libarchive version.
Update header checks to abort if not found.

== 3.0 Major Beta Release ==
Basic capability to list contents of an archive without extracting to disk.
Bump version and update command help text.
Update README with additional option details.
Change nftw() to depth-first scan to handle restoring directory permissions correctly.
When sorting cause directories to be sorted after files and in descending order of nesting level.
Take out stray printf().
Detect some DICOM formats and use BSC for DICOM data.
Overhaul documentation part #1
Detect and handle uncompressed PDF files using libbsc.
Force binary/text data detection for tar archives.
Get rid of unnecessary CLI option.
Add full pipeline mode check when archiving.
Update to PackJPG 2.5i.
Fixes crash with malformed Jpeg.
Fix issue #12.
Fix issue #13.
Create output directory with correct mode.
Fix the flow where pathname list is not sorted.
Fix ppmd decompression bug introduced in previous commit.
Reduce compression level for automatic pathname sorting.
Change to extraction directory only after opening archive.
Free PPMD buffer after compression, rather than caching.
Introduce new API in allocator to release buffer to OS.
Release LZMA buffers after use.
Drastically reduce memory consumption of PPMD8 in adaptive mode (Use lower max model order).
Detect AR archives and set the type.
Re-use a less common type code for AR.
Use Dispack generically for all executables and AR archives.
Move MSDOS COM single-byte magic number checks to last in the list.
Move advanced options flag into context structure.
Include dtd files as text type.
Update PackJPG to version 2.5h.
Fix missing bounds checking in Delta2.
Check harder with more strides in Delta2 for extreme compression levels.
Make LibArchive filter process buffer more generic.
Include explicit CLI flags for PackJPG and Dispack.
Avoid auto-selection of filters if advanced options are specified.
Add more robust checks for Jpeg and packJPG format files in filter routine.
Use case-insensitive checks for extension names.
Enable more features based on compression level, when archiving.
Do not use Libbsc for TIFFs. Not all TIFFs compress well with Libbsc.
Fix DEBUG-STATS build for Dispack.
Use adapt2 as default compression in archive mode.
Add more filter auto-selection by compression level in archive mode.
Replace odd stride lengths in Delta2 with standard numeric type lengths and improve performance.
Remove fast path exit to allow compressing headers and zero paddings via LZ4.
Use libbsc for AVI and MP4 files.
Use Libbsc for MP4 and FLAC files.
Change some rare file type codes to indicate some common types.
Use Libbsc for TIFF images.
Workaround for packJPG limitation.
Remove external Libbsc option.
Add forked and optimized copy of LGPL version of Libbsc.
Strip out Sort Transform from Libbsc copy.
Reduce Libbsc memory use.
Avoid redundant adler32 of data block in Libbsc.
Use Libbsc for DNA Sequence data instead of PPMD. Faster, better compression.
Fix pz extension handling for real.
Avoid Delta2,LZP for TIFF files. Negatively impacts compression.
Use libbsc/ppmd for BMP files.
Fix extension based hashing.
Do not append .pz extension to filenames already having it.
Some code formatting changes.
Fix PackJPG library usage. PackJPG interface doc is incomplete, ugh!
Handle the case where PackJPG expands the file rather than compressing.
Fix Dispack decoding.
Add Dispack filter with auto-detection of x86 executables in archive mode.
More elaborate magic header based detection of 32-bit and 64-bit x86 binaries.
Always use fast-mode LZ4 in Adaptive modes.
Optimize preprocessed compression and avoid a bunch of memory copies.
Fix a crash.
Add a few more file types.
More comments.
Fix fd leak.
Fix issues when handling Jpegs where packJPG borks.
Add fast handling of totally incompressible data (like Jpegs) in adaptive modes.
Add function to indicate totally incompressible data when archiving.
Reformat if statements in some places to reduce branching.
Enable auto-filtering of archive entries based on compression level.
Miscellaneous fixes.
Work in progress changes for packJPG encoding and decoding.
Enhance custom LibArchive filter functionlity.
Add basic framework for file type based filters during libarchive stage.
Add packJPG filter for Jpeg files (not active yet).
Directory format changes for clarity.
Add own implementation of archive entry extraction to allow custom filters.
Fix magic number check for endianness.
Structured handling of file types.
Handling of already compressed data based on compression algorithm.
Add a few more extension types.
Leverage file type detection(archiver) to improve compression performance.
Use detected file/data type(archiver) for Adaptive compression modes.
Update type flags and add more extensions.
Add file type detection based on magic values.
Add extension based file type detection and setting segment data type.
Use Bob Jenkins Minimal Perfect Hash to check for known extensions.
Use semaphore signaling and direct buffer copy for extraction.
Clear off private xattrs when extracting.
Enable pathname sorting only for high compression levels.
Replace slow pipe with direct memory copy for archive extraction.
Avoid using pipe during archive creation. Use semaphores and direct memory copy.
Use mmap to read from the pathlist file for performance.
Functionality to sort pathnames based on file extension and size.
More changes for archiving.
Allow multiple filenames on command line when archiving.

utils.c
/*
 * This file is a part of Pcompress, a chunked parallel multi-
 * algorithm lossless compression and decompression program.
 *
 * Copyright (C) 2012 Moinak Ghosh. All rights reserved.
 * Use is subject to license terms.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * moinakg@belenix.org, http://moinakg.wordpress.com/
 *      
 * This program includes partly-modified public domain source
 * code from the LZMA SDK: http://www.7-zip.org/sdk.html
 */

#include <sys/types.h>
#include <sys/param.h>
#include <libgen.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdio.h>
#include <errno.h>
#include <link.h>
#include <rabin_polynomial.h>
#include <skein.h>

#include "utils.h"

/*
 * Checksum properties
 */
static struct {
	char	*name;
	cksum_t	cksum_id;
	int	bytes;
} cksum_props[] = {
	{"CRC64",	CKSUM_CRC64,	8},
	{"SKEIN256",	CKSUM_SKEIN256,	32},
	{"SKEIN512",	CKSUM_SKEIN512,	64}
};

extern uint64_t lzma_crc64(const uint8_t *buf, size_t size, uint64_t crc);
extern uint64_t lzma_crc64_8bchk(const uint8_t *buf, size_t size,
	uint64_t crc, uint64_t *cnt);

void
err_exit(int show_errno, const char *format, ...)
{
	int err = errno;
	va_list args;

	va_start(args, format);
	vfprintf(stderr, format, args);
	va_end(args);

	if (show_errno)
		fprintf(stderr, "\nError: %s\n", strerror(err));
	exit(1);
}

/*
 * Fetch the command name that started the current process.
 * The returned string must be freed by the caller.
 */
const char *
get_execname(const char *argv0)
{
	char path[MAXPATHLEN];
	char apath[128];
	char *tmp1, *tmp2;
	pid_t pid;

	/* The easiest case: we are in linux */
	if (readlink("/proc/self/exe", path, MAXPATHLEN) != -1) {
		return (strdup(basename(path)));
	}

	/* Next easy case: Solaris/Illumos */
	pid = getpid();
	sprintf(apath, "/proc/%d/path/a.out", pid);
	if (readlink(apath, path, MAXPATHLEN) != -1) {
		return (strdup(basename(path)));
	}

	/* Oops... not in linux, not in Solaris no  guarantee */
	/* check if we have something like execve("foobar", NULL, NULL) */
	if (argv0 == NULL) {
		/* Give up */
		return (strdup("Unknown"));
	}

	tmp1 = strdup(argv0);
	tmp2 = strdup(basename(tmp1));
	free(tmp1);
	return (tmp2);
}

/*
 * Routines to parse a numeric string which can have the following suffixes:
 * k - Kilobyte
 * m - Megabyte
 * g - Gigabyte
 *
 * The number should fit in an ssize_t data type.
 * Numeric overflow is also checked. The routine parse_numeric() returns
 * 1 if there was a numeric overflow.
 */
static int
raise_by_multiplier(ssize_t *val, int mult, int power) {
	ssize_t result;

	while (power-- > 0) {
		result = *val * mult;
		if (result/mult != *val)
			return (1);
		*val = result;
	}
	return (0);
}

int
parse_numeric(ssize_t *val, const char *str)
{
	int i, ovr;
	char *mult;

	*val = strtoll(str, &mult, 0);
	if (*mult != '\0') {
		switch (*mult) {
		  case 'k':
		  case 'K':
			ovr = raise_by_multiplier(val, 1024, 1);
			break;
		  case 'm':
		  case 'M':
			ovr = raise_by_multiplier(val, 1024, 2);
			break;
		  case 'g':
		  case 'G':
			ovr = raise_by_multiplier(val, 1024, 3);
			break;
		  default:
			ovr = 2;
		}
	}

	return (ovr);
}

/*
 * Convert number of bytes into human readable format
 */
char *
bytes_to_size(uint64_t bytes)
{
	static char num[20];
	uint64_t kilobyte = 1024;
	uint64_t megabyte = kilobyte * 1024;
	uint64_t gigabyte = megabyte * 1024;
	uint64_t terabyte = gigabyte * 1024;

	if (bytes < kilobyte) {
		sprintf(num, "%llu B", bytes);

	} else if (bytes < megabyte) {
		sprintf(num, "%llu KB", bytes / kilobyte);

	} else if (bytes < gigabyte) {
		sprintf(num, "%llu MB", bytes / megabyte);

	} else if (bytes < terabyte) {
		sprintf(num, "%llu GB", bytes / gigabyte);

	} else {
		sprintf(num, "%llu B", bytes);
	}
	return (num);
}

/*
 * Read/Write helpers to ensure a full chunk is read or written
 * unless there is an error.
 * Additionally can be given an offset in the buf where the data
 * should be inserted.
 */
ssize_t
Read(int fd, void *buf, size_t count)
{
	ssize_t rcount, rem;
	uchar_t *cbuf;
	va_list args;

	rem = count;
	cbuf = (uchar_t *)buf;
	do {
		rcount = read(fd, cbuf, rem);
		if (rcount < 0) return (rcount);
		if (rcount == 0) break;
		rem = rem - rcount;
		cbuf += rcount;
	} while (rem);
	return (count - rem);
}

/*
 * Read the requested chunk and return the last rabin boundary in the chunk.
 * This helps in splitting chunks at rabin boundaries rather than fixed points.
 * The request buffer may have some data at the beginning carried over from
 * after the previous rabin boundary.
 */
ssize_t
Read_Adjusted(int fd, uchar_t *buf, size_t count, ssize_t *rabin_count, void *ctx)
{
        char *buf2;
        ssize_t rcount;
        rabin_context_t *rctx = (rabin_context_t *)ctx;

        if (!ctx)  return (Read(fd, buf, count));
        buf2 = buf;
        if (*rabin_count) {
                buf2 = (char *)buf + *rabin_count;
                count -= *rabin_count;
        }
        rcount = Read(fd, buf2, count);
        if (rcount > 0) {
                rcount += *rabin_count;
		if (rcount == count)
			rabin_dedup(rctx, buf, &rcount, 0, rabin_count);
		else
			*rabin_count = 0;
        } else {
                if (rcount == 0) rcount = *rabin_count;
                *rabin_count = 0;
        }
        return (rcount);
}

ssize_t
Write(int fd, const void *buf, size_t count)
{
	ssize_t wcount, rem;
	uchar_t *cbuf;

	rem = count;
	cbuf = (uchar_t *)buf;
	do {
		wcount = write(fd, cbuf, rem);
		if (wcount < 0) return (wcount);
		rem = rem - wcount;
		cbuf += wcount;
	} while (rem);
	return (count - rem);
}

/*
 * Thread sizing. We want a balanced combination of chunk threads and compression
 * algorithm threads that best fit the available/allowed number of processors.
 */
void
set_threadcounts(algo_props_t *props, int *nthreads, int nprocs, algo_threads_type_t typ) {
	int mt_capable;

	if (typ == COMPRESS_THREADS)
		mt_capable = props->compress_mt_capable;
	else
		mt_capable = props->decompress_mt_capable;

	if (mt_capable) {
		int nthreads1, p_max;

		if (nprocs == 3) {
			props->nthreads = 1;
			*nthreads = 3;
			return;
		}

		if (typ == COMPRESS_THREADS)
			p_max = props->c_max_threads;
		else
			p_max = props->d_max_threads;

		nthreads1 = 1;
		props->nthreads = 1;
		while (nthreads1 < *nthreads || props->nthreads < p_max) {
			if ((props->nthreads+1) * nthreads1 <= nprocs && props->nthreads < p_max) {
				props->nthreads++;

			} else if (props->nthreads * (nthreads1+1) <= nprocs && nthreads1 < *nthreads) {
				nthreads1++;
			} else {
				break;
			}
		}
		*nthreads = nthreads1;

	} else if (props->single_chunk_mt_capable && props->is_single_chunk) {
		*nthreads = 1;
		if (typ == COMPRESS_THREADS)
			props->nthreads = props->c_max_threads;
		else
			props->nthreads = props->d_max_threads;
		if (props->nthreads > nprocs)
			props->nthreads = nprocs;
	}
}

int
compute_checksum(uchar_t *cksum_buf, int cksum, uchar_t *buf, ssize_t bytes)
{
	if (cksum == CKSUM_CRC64) {
		uint64_t *ck = (uint64_t *)cksum_buf;
		*ck = lzma_crc64(buf, bytes, 0);

	} else if (cksum == CKSUM_SKEIN256) {
		Skein_512_Ctxt_t ctx;

		Skein_512_Init(&ctx, 256);
		Skein_512_Update(&ctx, buf, bytes);
		Skein_512_Final(&ctx, cksum_buf);

	} else if (cksum == CKSUM_SKEIN512) {
		Skein_512_Ctxt_t ctx;

		Skein_512_Init(&ctx, 512);
		Skein_512_Update(&ctx, buf, bytes);
		Skein_512_Final(&ctx, cksum_buf);
	} else {
		return (-1);
	}
	return (0);
}

/*
 * Check is either the given checksum name or id is valid and
 * return it's properties.
 */
int
get_checksum_props(char *name, int *cksum, int *cksum_bytes)
{
	int i;

	for (i=0; i<sizeof (cksum_props); i++) {
		if ((name != NULL && strcmp(name, cksum_props[i].name) == 0) ||
		    (*cksum != 0 && *cksum == cksum_props[i].cksum_id)) {
			*cksum = cksum_props[i].cksum_id;
			*cksum_bytes = cksum_props[i].bytes;
			return (0);
		}
	}
	return (-1);
}

/*
 * Endian independent way of storing the checksum bytes. This is actually
 * storing in little endian format and a copy can be avoided in x86 land.
 * However unsightly ifdefs are avoided here since this is not so performance
 * critical.
 */
void
serialize_checksum(uchar_t *checksum, uchar_t *buf, int cksum_bytes)
{
	int i,j;

	j = 0;
	for (i=cksum_bytes; i>0; i--) {
		buf[j] = checksum[i-1];
		j++;
	}
}

void
deserialize_checksum(uchar_t *checksum, uchar_t *buf, int cksum_bytes)
{
	int i,j;

	j = 0;
	for (i=cksum_bytes; i>0; i--) {
		checksum[i-1] = buf[j];
		j++;
	}
}
